core/drivers/crypto/hisilicon/sec_cipher.c

```diff
@@ -0,0 +1,774 @@
+		EMSG("Dupdate input param error");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	c_ctx = dupdate->ctx;
+	switch (c_ctx->alg) {
+	case C_ALG_SM4:
+	case C_ALG_AES:
+		if (sec_cipher_cryptlen_check(c_ctx, dupdate->src.length))
+			return TEE_ERROR_BAD_PARAMETERS;
+		break;
+	case C_ALG_DES:
+	case C_ALG_3DES:
+		if (dupdate->src.length % TEE_DES_BLOCK_SIZE) {
+			EMSG("Invalid src len");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sec_alloc_buffer(struct sec_cipher_ctx *c_ctx)
+{
+	c_ctx->in = malloc(c_ctx->len);
+	if (!c_ctx->in) {
+		EMSG("Fail to alloc c_in buf");
+		return TEE_ERROR_STORAGE_NO_SPACE;
+	}
+
+	c_ctx->in_dma = virt_to_phys(c_ctx->in);
+	if (!c_ctx->in_dma) {
+		EMSG("Fail to map c_in_dma");
+		goto free_c_in;
+	}
+
+	c_ctx->out = malloc(c_ctx->len);
+	if (!c_ctx->out) {
+		EMSG("Fail to alloc c_out buf");
+		goto free_c_in;
+	}
+
+	c_ctx->out_dma = virt_to_phys(c_ctx->out);
+	if (!c_ctx->out_dma) {
+		EMSG("Fail to map c_out_dma");
+		goto free_c_out;
+	}
+
+	return TEE_SUCCESS;
+
+free_c_out:
+	free(c_ctx->out);
+free_c_in:
+	free(c_ctx->in);
+	return TEE_ERROR_STORAGE_NO_SPACE;
+}
+
+static void sec_free_buffer(struct sec_cipher_ctx *c_ctx)
+{
+	free(c_ctx->in);
+	free(c_ctx->out);
+}
+
+static TEE_Result sec_cipher_update(struct drvcrypt_cipher_update *dupdate)
+{
+	struct sec_cipher_ctx *c_ctx = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+	size_t padding_size = 0;
+
+	ret = sec_cipher_param_check(dupdate);
+	if (ret)
+		return ret;
+
+	c_ctx = dupdate->ctx;
+	if (c_ctx->mode == C_MODE_CTR && (c_ctx->offs & CTR_SRC_ALIGN_MASK))
+		padding_size = c_ctx->offs % CTR_SRC_BLOCK_SIZE;
+
+	c_ctx->offs += dupdate->src.length;
+	c_ctx->len = dupdate->src.length + padding_size;
+	ret = sec_alloc_buffer(c_ctx);
+	if (ret)
+		return ret;
+
+	memset(c_ctx->in, 0, padding_size);
+	memcpy(c_ctx->in + padding_size, dupdate->src.data,
+	       dupdate->src.length);
+
+	ret = sec_do_cipher_task(c_ctx->qp, c_ctx);
+	if (ret)
+		goto free_buffer;
+
+	ret = sec_update_iv(c_ctx);
+	if (ret) {
+		EMSG("Fail to update iv, ret=%#"PRIx32, ret);
+		goto free_buffer;
+	}
+
+	memcpy(dupdate->dst.data, c_ctx->out + padding_size,
+	       dupdate->src.length);
+
+free_buffer:
+	sec_free_buffer(c_ctx);
+	return ret;
+}
+
+static void sec_cipher_final(void *ctx __unused)
+{
+}
+
+static void sec_cipher_copy_state(void *dst_ctx, void *src_ctx)
+{
+	struct sec_cipher_ctx *dst_c_ctx = dst_ctx;
+	struct sec_cipher_ctx *src_c_ctx = src_ctx;
+
+	dst_c_ctx->alg = src_c_ctx->alg;
+	dst_c_ctx->mode = src_c_ctx->mode;
+	dst_c_ctx->encrypt = src_c_ctx->encrypt;
+	dst_c_ctx->offs = src_c_ctx->offs;
+
+	if (src_c_ctx->key_len) {
+		dst_c_ctx->key_len = src_c_ctx->key_len;
+		dst_c_ctx->c_key_len = src_c_ctx->c_key_len;
+		memcpy(dst_c_ctx->key, src_c_ctx->key, dst_c_ctx->key_len);
+		dst_c_ctx->key_dma = virt_to_phys(dst_c_ctx->key);
+	}
```

Comment:
You may need an `else` here to initialize all fields with something.

Reply:
when key_len is 0，need to initialize all fields wilth zero？

Reply:
It's more safe to have it in a known state.